lib/Dialect/FIRRTL/FieldRefCache.cpp

```diff
@@ -0,0 +1,136 @@
+#include "circt/Dialect/FIRRTL/FIRRTLUtils.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/Support/FormatVariadic.h"
+
+using namespace circt;
+using namespace firrtl;
+
+FieldRef firrtl::FieldRefCache::getFieldRefFromValue(Value value,
+                                                     bool lookThroughCasts) {
+  using Info = std::pair<Value, size_t>;
+  // Vector of values and delta to next entry.
+  // Last will be root (delta == 0), so walking backwards constructs
+  // FieldRef's for all visited operations.
+  SmallVector<Info> indexing;
+  // Ignore null value for simplicity.
+  if (!value)
+    return FieldRef();
+
+#ifndef NDEBUG
+  ++queries;
+#endif
+
+  while (value) {
+    // Check cache to see if already visited.
+    auto it = refs.find(Key(value, lookThroughCasts));
+    if (it != refs.end()) {
+      // Found! If entire query is hit, we're done.
+#ifndef NDEBUG
+      ++hits;
+#endif
+      auto ref = it->second;
+      if (indexing.empty())
+        return ref;
+      // Otherwise, add entry for this using cached FieldRef,
+      // and add the FieldRef's value as last (root) entry.
+      indexing.emplace_back(value, ref.getFieldID());
+      indexing.emplace_back(ref.getValue(), 0);
+      break;
+    }
+#ifndef NDEBUG
+    ++computed;
+#endif
+
+    Operation *op = value.getDefiningOp();
+
+    // If this is a block argument, we are done.
+    if (!op) {
+      indexing.emplace_back(value, 0);
+      break;
+    }
+
+    auto [newValue, adj] =
+        TypeSwitch<Operation *, Info>(op)
+            .Case<RefCastOp, ConstCastOp, UninferredResetCastOp>([&](auto op) {
+              if (!lookThroughCasts)
+                return Info{};
+              return Info{op.getInput(), 0};
+            })
+            .Case<SubfieldOp, OpenSubfieldOp>([&](auto subfieldOp) {
+              typename decltype(subfieldOp)::InputType bundleType =
+                  subfieldOp.getInput().getType();
+              return Info{subfieldOp.getInput(),
+                          bundleType.getFieldID(subfieldOp.getFieldIndex())};
+            })
+            .Case<SubindexOp, OpenSubindexOp>([&](auto subindexOp) {
+              typename decltype(subindexOp)::InputType vecType =
+                  subindexOp.getInput().getType();
+              return Info{subindexOp.getInput(),
+                          vecType.getFieldID(subindexOp.getIndex())};
+            })
+            .Case<RefSubOp>([&](RefSubOp refSubOp) {
+              auto refInputType = refSubOp.getInput().getType();
+              size_t delta = FIRRTLTypeSwitch<FIRRTLBaseType, size_t>(
+                                 refInputType.getType())
+                                 .Case<FVectorType, BundleType>([&](auto type) {
+                                   return type.getFieldID(refSubOp.getIndex());
+                                 });
+              return Info{refSubOp.getInput(), delta};
+            })
+            .Default(Info{});
```

Comment:
Can we factor out this part into helper and share with plain `getFieldRefFromValue` implementation?

Reply:
On it, few minor refactoring + helper PR's.  Thanks for pushing for this!